### PR TITLE
Doip 2019

### PR DESCRIPTION
**Pull request by peckto**
Some changes to the DoIP Stack.
Still work in progress.
- [ ] fix TCP socket problem during discover with re-use port
- [ ] add support for DoIP version 2019

### CODE REVIEW

File: src/gallia/commands/discover/doip.py [hunk 1]
```diff
@@ -418,9 +422,12 @@ async def create_DoIP_conn(  # noqa: PLR0913
                 await conn.write_routing_activation_request(
                     RoutingActivationRequestTypes(routing_activation_type)
                 )
+            except OSError as e:
+                logger.error(f"Connection error: {e}; exit")
+                sys.exit(1)
             except Exception as e:  # TODO this probably is too broad
                 logger.warning(
-                    f"[🫨] Got me some good errors when it should be working (dis an infinite loop): {e}"
+                    f"[🫨] Got me some good errors when it should be working (dis an infinite loop): {repr(e)}"
```
Comment (rumpelsepp) [inline src/gallia/commands/discover/doip.py]:
```
{repr(e)} -> {e!r}
```

File: src/gallia/transports/doip.py [hunk 2]
```diff
@@ -524,7 +548,7 @@ async def write_request_raw(self, hdr: GenericHeader, payload: DoIPOutData) -> N
 
     async def write_diag_request(self, data: bytes) -> None:
         hdr = GenericHeader(
-            ProtocolVersion=ProtocolVersions.ISO_13400_2_2012,
+            ProtocolVersion=ProtocolVersions.ISO_13400_2_2019,
```
Comment (rumpelsepp) [inline src/gallia/transports/doip.py:555]: Should this be configurable?

File: src/gallia/transports/doip.py [hunk 3]
```diff
@@ -540,7 +564,7 @@ async def write_routing_activation_request(
         activation_type: int,
     ) -> None:
         hdr = GenericHeader(
-            ProtocolVersion=ProtocolVersions.ISO_13400_2_2012,
```
Comment (rumpelsepp) [inline src/gallia/transports/doip.py:543]: Should this be configurable?

File: src/gallia/transports/doip.py [hunk 4]
```diff
@@ -553,7 +577,7 @@ async def write_routing_activation_request(
 
     async def write_alive_check_response(self) -> None:
         hdr = GenericHeader(
-            ProtocolVersion=ProtocolVersions.ISO_13400_2_2012,
```
Comment (rumpelsepp) [inline src/gallia/transports/doip.py:556]: Should this be configurable?

File: src/gallia/transports/doip.py [hunk 5]
```diff
@@ -302,6 +316,7 @@ def pack(self) -> bytes:
     | DiagnosticMessagePositiveAcknowledgement
     | DiagnosticMessageNegativeAcknowledgement
     | AliveCheckRequest
+    | DoIPNegativeAckResponse
```
Comment (rumpelsepp) [inline src/gallia/transports/doip.py:320]: Is the name of this reponse `DoiPNegativeAckResponse` or `NegativeAckResponse`? The naming is inconsistent. What does the standard say?

File: src/gallia/transports/doip.py [hunk 6]
```diff
@@ -182,6 +184,18 @@ def pack(self) -> bytes:
         )
 
 
+@dataclass
+class DoIPNegativeAckResponse:
+    nack_code: DiagnosticMessageNegativeAckCodes
+
+    @classmethod
+    def unpack(cls, data: bytes) -> DoIPNegativeAckResponse:
```
Comment (rumpelsepp) [inline src/gallia/transports/doip.py:193]: we could return `typing.Self` here.

**Review by rumpelsepp — CHANGES_REQUESTED** [review]

**Comment by ferdinandjarisch** [comment]
I've moved my fixes regarding general DoIP discovery behavior and especially the issue with stalling scans to #510, since they are independent of the DoIP 2019 header.

After they have been reviewed and merged, this PR should be carefully rebased, such that only DoIP 2019-relevant parts of 445ab962d6e48b3fe57f129b8cdd88d9fafee026 remain.

I then have further comments about implementing support for the 2019 standard.

**Comment by ferdinandjarisch** [comment]
Surpassed by #510

**Pull request closed without merging**